rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/action/TbDeviceStateNode.java

```diff
@@ -0,0 +1,194 @@
+                " Incoming message is forwarded using the <code>Success</code> chain," +
+                " unless an unexpected error occurs during message processing" +
+                " then incoming message is forwarded using the <code>Failure</code> chain." +
+                "<br>" +
+                "Supported device connectivity events are:" +
+                "<ul>" +
+                "<li>Connect event</li>" +
+                "<li>Disconnect event</li>" +
+                "<li>Activity event</li>" +
+                "<li>Inactivity event</li>" +
+                "</ul>" +
+                "This node is particularly useful when device isn't using transports to receive data," +
+                " such as when fetching data from external API or computing new data within the rule chain.",
+        configClazz = TbDeviceStateNodeConfiguration.class,
+        uiResources = {"static/rulenode/rulenode-core-config.js"},
+        configDirective = "tbActionNodeDeviceStateConfig"
+)
+public class TbDeviceStateNode implements TbNode {
+
+    static final Set<TbMsgType> SUPPORTED_EVENTS = EnumSet.of(
+            TbMsgType.CONNECT_EVENT, TbMsgType.ACTIVITY_EVENT, TbMsgType.DISCONNECT_EVENT, TbMsgType.INACTIVITY_EVENT
+    );
+
+    private TbMsgType event;
+
+    @Override
+    public void init(TbContext ctx, TbNodeConfiguration configuration) throws TbNodeException {
+        TbMsgType event = TbNodeUtils.convert(configuration, TbDeviceStateNodeConfiguration.class).getEvent();
+        if (event == null) {
+            throw new TbNodeException("Event cannot be null!", true);
+        }
+        if (!SUPPORTED_EVENTS.contains(event)) {
+            throw new TbNodeException("Unsupported event: " + event, true);
+        }
+        this.event = event;
+    }
+
+    @Override
+    public void onMsg(TbContext ctx, TbMsg msg) {
+        var originator = msg.getOriginator();
+        if (!ctx.isLocalEntity(originator)) {
+            log.warn("[{}][{}] Received message from non-local entity [{}]!",
+                    ctx.getTenantId().getId(), ctx.getSelfId().getId(), originator.getId());
+            ctx.ack(msg);
+            return;
+        }
+        if (!EntityType.DEVICE.equals(originator.getEntityType())) {
+            ctx.tellSuccess(msg);
+            return;
+        }
+        switch (event) {
+            case CONNECT_EVENT: {
+                sendDeviceConnectMsg(ctx, msg);
+                break;
+            }
+            case ACTIVITY_EVENT: {
+                sendDeviceActivityMsg(ctx, msg);
+                break;
+            }
+            case DISCONNECT_EVENT: {
+                sendDeviceDisconnectMsg(ctx, msg);
+                break;
+            }
+            case INACTIVITY_EVENT: {
+                sendDeviceInactivityMsg(ctx, msg);
+                break;
+            }
+            default: {
+                ctx.tellFailure(msg, new IllegalStateException("Configured event [" + event + "] is not supported!"));
```

Comment:
ctx.tellSuccess is missing for the "normal" cases. This will cause processing timeouts. Please fix this.